Fixed bugs and refactor of main file.

HerissonMignion opened the pull request:
Fixed bugs related to missing quotes when calling functions or programs. Refactor the main file and the argument parsing. Because arguments are parsed slightly differently, i've also rewritten the help message and adjusted the arise command in the github workflows.

HerissonMignion committed:
fixed many bugs related to missing quotes when calling functions/programs. overhauled the main file/entry point and the help message accordingly.

HerissonMignion committed:
added missing quotes in one more file

neonspectra commented on arise:
```diff
@@ -10,14 +14,27 @@ arise_version="1.1.0"
 
-cd "$(dirname $0)"
+cd "$SCRIPT_DIR"
 
```
```suggestion

# Display help if no arguments are provided
if [ "$#" -eq 0 ]; then
    arise_help
    exit 0
fi

# Ensure first argument is 'build' for default behaviour. Shift arguments and continue. 
if [ "$1" != "build" ]; then
    arise_help
    exit 1
fi
shift

```

neonspectra replied:
Maintains existing invocation to avoid API breaking change. Potentially a little ugly, but keeps showing the help as the default invocation action.

The reason I chose this invocation format in the first place is because I think that even though calling the program with no arguments is cleaner, it's friendlier to display help information with a no-arguments call than to just start building away. This way someone who doesn't know what they're doing is greeted with some help info if they try to call the program with no arguments. 

Granted, this is should not matter much since invocations of this program will generally happen via CI calls so to invocation line isn't that important for a program like this either way.

neonspectra commented on .github/workflows/ci-xml-reserved-characters.yml:
```diff
@@ -33,7 +33,7 @@ jobs:
 
       - name: Build Arise
-        run: bash arise build
+        run: bash arise
```
rejecting breaking change to invocation syntax
```suggestion
        run: bash arise build
```

neonspectra commented on .github/workflows/arise-deploy.yml:
```diff
@@ -44,7 +44,7 @@ jobs:
 
       - name: Build Arise
-        run: bash arise build
+        run: bash arise
```
rejecting breaking change to invocation syntax
```suggestion
        run: bash arise build
```

neonspectra commented on arise:
```diff
@@ -10,14 +14,27 @@ arise_version="1.1.0"
 ##############################################################
 
-cd "$(dirname $0)"
+cd "$SCRIPT_DIR"
```
Set the default as an option so that a user can change the SCRIPT_DIR if they want to above; otherwise they leave it unset and it uses the default.

```suggestion
# Default to the script's directory if SCRIPT_DIR is not set
# https://stackoverflow.com/questions/59895/how-do-i-get-the-directory-where-a-bash-script-is-located-from-within-the-script
SCRIPT_DIR_DEFAULT=$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )

# Check if SCRIPT_DIR is set and non-empty using parameter expansion; otherwise use SCRIPT_DIR_DEFAULT
cd "${SCRIPT_DIR:-$SCRIPT_DIR_DEFAULT}" || {
    echo "Error: Unable to change directory to '${SCRIPT_DIR:-$SCRIPT_DIR_DEFAULT}'"
    exit 1
}
```

HerissonMignion replied:
I think it's nice for the moment (and the forseable future?) that arise is small enough that it comes with the website that it builds, kind of like the way the gnu autotools are "installed" in the project they compile (but without the complexity of gnu ~~autohell~~ autotools). (EDIT: And therefore I dont see the point of making this configurable.)

But if you insist, I'll do as you say.

neonspectra commented on ci/xml-reserved-characters.sh:
```diff
@@ -10,7 +10,7 @@ echo ""
 
 echo "Testing to ensure the Arise site built the test suite page..."
-if [ -f $testpage ] 
+if [ -f "$testpage"" ] 
```
remove extra quote
```suggestion
if [ -f "$testpage" ] 
```

neonspectra commented on lib/functions/inline/arise_help.sh:
```diff
@@ -9,29 +9,63 @@
+
+SYNOPSIS
+
+	arise [<options>]...
+
+	Options must be given separately (e.g. "arise -f -k".)
```
see above comments regarding keeping existing invocation syntax.
```suggestion
	arise build [<options>]...

	Options must be given separately (e.g. "arise build -f -k".)
```

neonspectra commented on lib/functions/subshell/build_rss.sh:
```diff
@@ -31,34 +31,35 @@ cat > $rss <<EOF
-                        
-                        cat >> $rss <<EOF
+				if [[ $rss_hide != "true" ]] && [[ $is_toc != "true" ]]; then
+					# Convert html's ISO8601 date to RSS's RFC-822. stupid RSS.
```
look sometimes i just want to employ sentence enhancers for a little whimsy in my comments so that the reader can share in my deranged grievances. compromise? ;)
```suggestion
					# Convert html's ISO8601 date to RSS's RFC-822. sometimes you need a stupid hack to fix an even stupider RFC decision 😐
```

neonspectra commented on lib/functions/inline/arise_logo.sh:
```diff
@@ -9,7 +9,7 @@
 
 arise_logo() {
-cat <<EOF
+	cat <<"EOF"
```
no reason to enable variable expansion here
```suggestion
	cat <<EOF
```

HerissonMignion replied:
putting it in quotes is what disables variable expansion

neonspectra commented on arise:
```diff
@@ -1,4 +1,8 @@
 #!/bin/bash
+
+# https://stackoverflow.com/questions/59895/how-do-i-get-the-directory-where-a-bash-script-is-located-from-within-the-script
+SCRIPT_DIR=$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )
```
I think the reason you modified this and moved it up here is to allow a config option to make it easier to overwrite To make this cleaner, I changed it to an optional item; see additional suggestion below to use the snippet by default but allow overwrite. 

Since PR base branch changes aren't disabled and GitHub reviews won't let me comment on the entire file, can I ask that you move this option below the version tag stub but above the 'begin main function' stub?

```suggestion
# If you want to manually specify the path to the arise binary, you can uncomment and modify the variable below.
# This should not be necessary in most cases and this should generally be left as default.
# SCRIPT_DIR="/change/this/to/path/to/arise"
```

HerissonMignion replied:
The reason I added this is because the script(s) is very reliant on relative paths to the root of the arise folder. My rationale is that it doesn't hurt to begin with a cd to the directory of the arise (and I get the path via this line) instead of relying on the user or external factors to have the appropriate current directory.

The reason this is at the top of the file is because that stackoverflow answer that I always copy paste in such situation says to be aware that this must be done before cd-ing to any other directory, so i paste it on the first lines.

neonspectra reviewed (changes requested):
Hey! Thank you for the large PR! Sorry for taking some time to review it; things are hectic for me so tech hobby stuff has been on the backburner for me.

I've worked my way through your contributions and requested some changes. Some thoughts, as well as comments since it appears that PR base branch modification is disabled:
- I requested changes to keep the `arise build` syntax. More comments on this in the suggestions.
    - The main script refactor is really nice. It does technically introduce breaking changes since you can't mix args together with your implementation, but your argument handling is ultimately a lot cleaner than the hacky way I did it.
    - I tried to keep your changes while also preserving syntax that would most definitely break someone's deployment if pulling in a new main without being careful about it.
- I really appreciate your rewrite of the help snippet. Can you update `docs/guides/running-arise-locally` to either reference the new information, or cut the old help snippet in there and simply request that the user reference `/lib/functions/inline/arise_help.sh` for the full information? 
    - Doesn't matter to me which way; but i think it makes sense to update this doc as part of merging a PR with changes to the help syntax.
- TY for the quote linting work. I really need to set up shellcheck on this repo at some point but it's low on my to-do list since this project works well enough for my personal website I'm using it for.

Hope these comments all make sense. Thanks again for your contribution!

HerissonMignion committed:
Update arise

Co-authored-by: Neon <10284121+neonspectra@users.noreply.github.com>

HerissonMignion committed:
Update .github/workflows/ci-xml-reserved-characters.yml

Co-authored-by: Neon <10284121+neonspectra@users.noreply.github.com>

HerissonMignion committed:
Update .github/workflows/arise-deploy.yml

Co-authored-by: Neon <10284121+neonspectra@users.noreply.github.com>

HerissonMignion committed:
Update ci/xml-reserved-characters.sh

Co-authored-by: Neon <10284121+neonspectra@users.noreply.github.com>

HerissonMignion committed:
Update lib/functions/inline/arise_help.sh

Co-authored-by: Neon <10284121+neonspectra@users.noreply.github.com>

HerissonMignion committed:
Update lib/functions/subshell/build_rss.sh

Co-authored-by: Neon <10284121+neonspectra@users.noreply.github.com>

HerissonMignion committed:
moved suggestion to appropriate location

HerissonMignion committed:
fixed argument parsed to support combined options (-asdf)

HerissonMignion committed:
fixed doc

HerissonMignion committed:
switched space to tab

HerissonMignion commented:
Well, I think I addressed most of your concerns. I implemented support for combined options (ex: -sf) in argument parsing so that it's more compatible with previous implementation.

Just know that i did not take the time to test that CI/CD related stuff "still works". I think "they still work", but because of the "simple" nature of the changes of this PR i'm too lazy to go out of my way to test that forking the repo and following the instructions still work. I'm just saying.
